Add new drop mode ON_GRID for GridDropTarget

pleku opened the pull request:
Also adds a way to not accept drops on rows when the user has sorted the grid.
This way the bad UX can be avoided for showing the drop indicator for the wrong place when the grid has been sorted.
This has not been made default behavior to GridDropTarget since it would change behavior compared to 8.1.
Instead if is triggerable via API in GridDropTarget.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/10296)
<!-- Reviewable:end -->

Commit:
Add new drop mode ON_GRID for GridDropTarget

Also adds a way to not accept drops on rows when the user has sorted the grid.
This way the bad UX can be avoided for showing the drop indicator for the wrong place when the grid has been sorted.
This has not been made default behavior to GridDropTarget since it would change behavior compared to 8.1.
Instead if is triggerable via API in GridDropTarget.

pleku commented:
This is a prequisite for #10255

pekam commented:
Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQraFlNDzdNVfONjH8:-KyQraFlNDzdNVfONjH9:b-kd5etl) (no related file):*
Javadocs for methods in `com.vaadin.ui.components.grid.GridDropEvent` could be updated to mention what happens to `dropTargetRow` and `dropLocation` when the drop mode is `ON_GRID`.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR-fnEdrUmmsLRp-kf:-KyR-fnFSgZ4i-Nu96On:b-y77nqj) (no related file):*
So setting `dropTarget.setDropAllowedOnSortedGridRows(false);` and sorting the grid causes the grid to behave like the drop mode would be `ON_GRID`, but it doesn't actually change the value of drop mode.

So in this case `dropTarget.getDropMode()` gives incorrect information about the drop target behavior.

Wouldn't it be better to just change to `DropMode.ON_GRID` in case of `is drop disabled on sorted rows && is grid sorted` and change back to the user-defined drop mode when that condition doesn't hold anymore?

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 83 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQiIRR0uqPVhfCPp06:-KyQiIRR0uqPVhfCPp07:bpltm2) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L83)):*
> ```Java
>     /**
>      * Class name to apply when dragged over an empty grid, or when dropping on
>      * rows it not possible (see {@link #isDroppingOnRowsPossible()}).
> ```

"when dropping on rows **is** not possible"

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 109 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQgt66ycb7zbq6LjdL:-KyQgt66ycb7zbq6LjdM:b-f3b9rw) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L109)):*
> ```Java
>      * empty.
>      *
>      * @return {@code} true when the drop target is the whole grid, or
> ```

`{@code true}`

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 110 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQh0kjiauiMOmz-JFe:-KyQh0kjiauiMOmz-JFf:byu88wc) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L110)):*
> ```Java
>      *
>      * @return {@code} true when the drop target is the whole grid, or
>      *         {@code false} when it is the one of the rows
> ```

remove extra "the"

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 121 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQlvVCDPi5_c0rrAlg:-KyQlvVCDPi5_c0rrAlh:brgmuk1) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L121)):*
> ```Java
>         }
> 
>         if (getEscalator().getVisibleRowRange().isEmpty()) {
> ```

Is this the best way to find out if the grid is empty? Isn't it a bit hacky?

The logic will fail if someone creates a grid that allows scrolling down below all the rows so that none are visible. Or maybe even if the height of the grid is zero?

On the other hand, if it's enough to check if there are **visible** rows, you could just edit the javadoc.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 132 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQvxNY13_5zbbZu2d0:-KyQvxNY13_5zbbZu2d1:b-pthtob) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L132)):*
> ```Java
>             Map<String, String> data, String dropEffect,
>             NativeEvent dropEvent) {
>         // the next on is always either the table wrapper or a row element
> ```

"the next on"? You mean "one"? Why changed from the original "the target element"?

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java, line 50 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQhQ8hSQxUTM_CPnm6:-KyQhQ8iwJR7_M_mJnuP:bc5bwpi) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java#L50)):*
> ```Java
>      * </ul>
>      */
>     ON_TOP_OR_BETWEEN,
> ```

An empty line after this would make the style consistent. Not blocking.

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java, line 44 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQhmh0MnWrjzBkXang:-KyQhmh0MnWrjzBkXanh:b-f3b9rw) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java#L44)):*
> ```Java
>      * Can the user drop on grid rows if the grid is sorted.
>      * <p>
>      * Default value is {@code} true for backwards compatibility with 8.1.
> ```

`{@code true}`

---

*[uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java, line 94 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR3cMG148bT1gs86nC:-KyR3cMG148bT1gs86nD:b-x8u201) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java#L94)):*
> ```Java
>         });
> 
>         CheckBox dropOnSortedGridRows = new CheckBox("Drop on Sorted Grid Rows",
> ```

You have just copy-pasted the code above, with minor changes? How does this test anything?

---

*[uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java, line 96 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR4-lDe-3ryHlQmutx:-KyR4-lDe-3ryHlQmuty:b86i3ns) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java#L96)):*
> ```Java
>         CheckBox dropOnSortedGridRows = new CheckBox("Drop on Sorted Grid Rows",
>                 dropTarget.isDropAllowedOnSortedGridRows());
>         transitionCheckBox.addValueChangeListener(event -> {
> ```

You forgot to rename this checkbox, but it looks like the test is WIP anyway

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296)*
<!-- Sent from Reviewable.io -->

pekam commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions.

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java, line 115 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBRHponHLcgC0qWST:-KyRBRHponHLcgC0qWSU:baenriu) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java#L115)):*
> ```Java
>      *            {@code true} for allowing, {@code false} for not allowing
>      *            drops on sorted grid rows
>      * @since
> ```

no version defined for `@since`

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java, line 131 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBJAVsl0m1QkU9yNo:-KyRBJAVsl0m1QkU9yNp:baenriu) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java#L131)):*
> ```Java
>      * @return whether drop are allowed for the grid's rows when user has sorted
>      *         the grid
>      * @since
> ```

no version defined for `@since`

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java, line 58 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBViHk6WiRqcPFvgW:-KyRBViHk6WiRqcPFvgX:baenriu) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java#L58)):*
> ```Java
>      * to be used automatically when the user has sorted the grid.
>      *
>      * @since
> ```

no version defined for `@since`

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java, line 46 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBXD7SGuA3KZg26TT:-KyRBXD7SGuA3KZg26TU:baenriu) ([raw file](https://github.com/vaadin/framework/blob/62bd060c37b8cf32e1d01abac74ef3bf0aa4ff92/shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java#L46)):*
> ```Java
>      * Default value is {@code} true for backwards compatibility with 8.1.
>      *
>      * @since
> ```

no version defined for `@since`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296)*
<!-- Sent from Reviewable.io -->

Commit:
Refactor sorted grid drop logic to server side

pleku commented:
Review status: 2 of 9 files reviewed at latest revision, 14 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQraFlNDzdNVfONjH8:-KyqdYBiHM_eMF9zxD2V:bb6un1j) (no related file):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

Javadocs for methods in `com.vaadin.ui.components.grid.GridDropEvent` could be updated to mention what happens to `dropTargetRow` and `dropLocation` when the drop mode is `ON_GRID`.
</blockquote></details>

Done

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR-fnEdrUmmsLRp-kf:-KyqkCVNPhbT1AhKPG8q:bf7lamc) (no related file):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

So setting `dropTarget.setDropAllowedOnSortedGridRows(false);` and sorting the grid causes the grid to behave like the drop mode would be `ON_GRID`, but it doesn't actually change the value of drop mode.

So in this case `dropTarget.getDropMode()` gives incorrect information about the drop target behavior.

Wouldn't it be better to just change to `DropMode.ON_GRID` in case of `is drop disabled on sorted rows && is grid sorted` and change back to the user-defined drop mode when that condition doesn't hold anymore?
</blockquote></details>

Done, had forgotten that there was sort listeners for grid.

I actually think doing this made the code more complex in general, since the server side part for handling that is not pretty with all the special cases.  It only reduced one if check from client side and the shared state variable.

I'll let @tsuoanttila be the judge.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 83 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQiIRR0uqPVhfCPp06:-KyqklW28rcZLpRYPoy6:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L83)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

"when dropping on rows **is** not possible"
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 109 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQgt66ycb7zbq6LjdL:-Kyql1tDs_tQIj4cypYo:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L109)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

`{@code true}`
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 110 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQh0kjiauiMOmz-JFe:-Kyqp7uw6sLXwwg1WeH4:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L110)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

remove extra "the"
</blockquote></details>

Done.

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 121 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQlvVCDPi5_c0rrAlg:-KyqlGHTnNxcOOaqSjI6:b3iebiy) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L121)):*
> someone creates a grid that allows scrolling down below all the rows so that none are visible

I don't think that is possible, and if it was, let me know how can you drop on rows if none are visible ?

Or if the grid height is zero, how can one drop on top of a row ??? Actually if the height is zero, this will be NOOP since you'll never get a drop event from the browser for something that is not there....

---

*[client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java, line 132 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQvxNY13_5zbbZu2d0:-KyqmICxJTFOfXM6mxVM:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/client/src/main/java/com/vaadin/client/connectors/grid/GridDropTargetConnector.java#L132)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

"the next on"? You mean "one"? Why changed from the original "the target element"?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java, line 115 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBRHponHLcgC0qWST:-KyqmJ61miqTYkyS4S2_:bp4peb0) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java#L115)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

no version defined for `@since`
</blockquote></details>

That is the used practice, the sinces are updated once it is actually picked to a new version release. Part of release check list. And is handled by an automatic script.

It actually happened with this commit originally that the user put sinces for 8.1 and this was never released for that...

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java, line 131 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBJAVsl0m1QkU9yNo:-KyqmTIIoal1_QFoNR6V:b-cd2dtl) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/server/src/main/java/com/vaadin/ui/components/grid/GridDropTarget.java#L131)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

no version defined for `@since`
</blockquote></details>

Not needed

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java, line 58 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBViHk6WiRqcPFvgW:-KyqmcKcobAFxbJg2HdS:b-cd2dtl) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/shared/src/main/java/com/vaadin/shared/ui/grid/DropMode.java#L58)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

no version defined for `@since`
</blockquote></details>

Not needed

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java, line 44 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQhmh0MnWrjzBkXang:-Kyqml0TlnQxnSBM4lLL:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java#L44)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

`{@code true}`
</blockquote></details>

Done.

---

*[shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java, line 46 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyRBXD7SGuA3KZg26TT:-KyqmlHlOD8XCVylbjM2:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/shared/src/main/java/com/vaadin/shared/ui/grid/GridDropTargetState.java#L46)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

no version defined for `@since`
</blockquote></details>

Done.

---

*[uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java, line 94 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR3cMG148bT1gs86nC:-KyqnBjq-aO-bFSjfzHA:b-twkxp2) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java#L94)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

You have just copy-pasted the code above, with minor changes? How does this test anything?
</blockquote></details>

Oops, forgot to add latest

---

*[uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java, line 96 at r1](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR4-lDe-3ryHlQmutx:-KyqnFqM8SGFvi4AQvpK:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d33059bd68845305aa50aa7342335f64a063139e/uitest/src/main/java/com/vaadin/tests/components/grid/GridDragAndDrop.java#L96)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

You forgot to rename this checkbox, but it looks like the test is WIP anyway
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296)*
<!-- Sent from Reviewable.io -->

pekam commented:
Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQraFlNDzdNVfONjH8:-Kyubscz7ZXQcpH3kCFi:b-yht06m) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Constructor still says only that `dropTargetRow == null` if dropped on empty grid

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR-fnEdrUmmsLRp-kf:-KyuqxiH8MnMe9ODisHA:b-b2op98) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Done, had forgotten that there was sort listeners for grid.

I actually think doing this made the code more complex in general, since the server side part for handling that is not pretty with all the special cases.  It only reduced one if check from client side and the shared state variable.

I'll let @tsuoanttila be the judge.
</blockquote></details>

Does it matter that it's more complicated if it's better for the user?

So I like it, but you didn't consider the case where the user changes the drop mode while it's automatically set to `ON_GRID`:
```
dropTarget.setDropMode(DropMode.BETWEEN);
// drop mode is BETWEEN

dropTarget.setDropAllowedOnSortedGridRows(false);
grid.sort();
// drop mode is ON_GRID

dropTarget.setDropMode(DropMode.ON_TOP);
// drop mode is ON_TOP (overrides the automatically set ON_GRID)

dropTarget.setDropAllowedOnSortedGridRows(true);
// drop mode is BETWEEN because that was cached, but it should be ON_TOP
```

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropEvent.java, line 92 at r2](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyubL-Eyc0Gq1wU84jM:-KyubL-Eyc0Gq1wU84jN:b-8txcwn) ([raw file](https://github.com/vaadin/framework/blob/c18bcfa579c84c07102d02e8b3b5a833f8bbad7b/server/src/main/java/com/vaadin/ui/components/grid/GridDropEvent.java#L92)):*
> ```Java
> 
>     /**
>      * Get the location of the drop within the row-
> ```

a random hyphen appeared

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296)*
<!-- Sent from Reviewable.io -->

Commit:
Block setDropMode calls

Blocking setDropMode set values if the grid has been sorted and drop on
sorted rows is not allowed. The value is used once the grid is not sorted
anymore or the drops are allowed on sorted rows.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/vaadin/framework?pullRequest=10296) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/vaadin/framework?pullRequest=10296) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/vaadin/framework?pullRequest=10296) it.</sub>

pleku commented:
Review status: 6 of 9 files reviewed at latest revision, 3 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyQraFlNDzdNVfONjH8:-KyvmLWHmmGWPqQ27HGt:bb6un1j) (no related file):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

Constructor still says only that `dropTargetRow == null` if dropped on empty grid
</blockquote></details>

Done

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR-fnEdrUmmsLRp-kf:-KyveAxkzT9lYNjTOaC6:blc6iaj) (no related file):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

Does it matter that it's more complicated if it's better for the user?

So I like it, but you didn't consider the case where the user changes the drop mode while it's automatically set to `ON_GRID`:
```
dropTarget.setDropMode(DropMode.BETWEEN);
// drop mode is BETWEEN

dropTarget.setDropAllowedOnSortedGridRows(false);
grid.sort();
// drop mode is ON_GRID

dropTarget.setDropMode(DropMode.ON_TOP);
// drop mode is ON_TOP (overrides the automatically set ON_GRID)

dropTarget.setDropAllowedOnSortedGridRows(true);
// drop mode is BETWEEN because that was cached, but it should be ON_TOP
```
</blockquote></details>

Very good point, that would actually lead to inconsistent state with between the `DropMode` and the `dropAllowedOnGridRows`. There are couple options for what to do
1) capture the set mode to the cachedDropMode and don't use it until grid sorting has been cleared or dropping on sorted rows is allowed
2) take the mode into use and set the drop allowed flag back to true, regardless if the grid has been sorted currently or not
3) take the mode into use and set the drop allowed flag back to true, but only for the latter if the grid is sorted
4) throw

1) Feels OK since you expect it to be used again when the grid is not sorted. It has the downside though that you use the setter and it might not be effective immediately if the grid has been sorted, but then you have touched the `dropAllowedOnGridRows` anyway. Thus after calling the setter the getter is not returning what you just set, but IMO that is just academical excuse to not do it like this. And I feel that once you've touched the `dropAllowedOnSortedGridRows` you know what you're doing.
2) is a side-effect, mutating two states at once is generally not desired, but it would be maybe less offending academically to switch the flag at that point.
3) is same as 2) but would allow to still keep the sorting option
4) would at least be explicit and understandable in the end, but might be a bit aggressive and cause WTF in the beginning, could lead into ugly code switching back values back and forth in some use cases.

I decided to go with 1) for now since it would allow you to swap the mode freely, but still not allow drops on rows if the grid is sorted. So by using the `dropAllowedOnSortedGridRows` you know that any set `dropMode` is not used when the grid is sorted, even if you change the mode. Which was actually already stated in the javadocs.

---

*[server/src/main/java/com/vaadin/ui/components/grid/GridDropEvent.java, line 92 at r2](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyubL-Eyc0Gq1wU84jM:-KyvfZppQdG794p_yO26:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/c18bcfa579c84c07102d02e8b3b5a833f8bbad7b/server/src/main/java/com/vaadin/ui/components/grid/GridDropEvent.java#L92)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

a random hyphen appeared
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296)*
<!-- Sent from Reviewable.io -->

pekam commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/vaadin/framework/10296#-KyR-fnEdrUmmsLRp-kf:-Kyw2CWds58RV61IYRns:buopzua) (no related file):*
<details><summary><i>Previously, pleku (Pekka Hyvönen) wrote…</i></summary><blockquote>

Very good point, that would actually lead to inconsistent state with between the `DropMode` and the `dropAllowedOnGridRows`. There are couple options for what to do
1) capture the set mode to the cachedDropMode and don't use it until grid sorting has been cleared or dropping on sorted rows is allowed
2) take the mode into use and set the drop allowed flag back to true, regardless if the grid has been sorted currently or not
3) take the mode into use and set the drop allowed flag back to true, but only for the latter if the grid is sorted
4) throw

1) Feels OK since you expect it to be used again when the grid is not sorted. It has the downside though that you use the setter and it might not be effective immediately if the grid has been sorted, but then you have touched the `dropAllowedOnGridRows` anyway. Thus after calling the setter the getter is not returning what you just set, but IMO that is just academical excuse to not do it like this. And I feel that once you've touched the `dropAllowedOnSortedGridRows` you know what you're doing.
2) is a side-effect, mutating two states at once is generally not desired, but it would be maybe less offending academically to switch the flag at that point.
3) is same as 2) but would allow to still keep the sorting option
4) would at least be explicit and understandable in the end, but might be a bit aggressive and cause WTF in the beginning, could lead into ugly code switching back values back and forth in some use cases.

I decided to go with 1) for now since it would allow you to swap the mode freely, but still not allow drops on rows if the grid is sorted. So by using the `dropAllowedOnSortedGridRows` you know that any set `dropMode` is not used when the grid is sorted, even if you change the mode. Which was actually already stated in the javadocs.


</blockquote></details>

I agree that's the way to go here. Do you still want to hear @tsuoanttila's opinion before merging?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/10296#-:-Kyw2tpkHqNgLRut0our:bnfp4nl)*
<!-- Sent from Reviewable.io -->
